src/python/dxpy/app_builder.py

```diff
@@ -212,58 +161,199 @@ def upload_resources(src_dir, project=None, folder='/', ensure_upload=False):
     if os.path.exists(resources_dir) and len(os.listdir(resources_dir)) > 0:
         target_folder = applet_spec['folder'] if 'folder' in applet_spec else folder
 
-        # Optimistically look for a resource bundle with the same
-        # contents, and reuse it if possible. The resource bundle
-        # carries a property 'resource_bundle_checksum' that indicates
-        # the checksum; the way in which the checksum is computed is
-        # given in the documentation of _directory_checksum.
+        # Copy the code previously in _directory_checksum here and we will create
+        # the tar file while performing the checksum.  If the checksum matches 
+        # (and ensure_upload is False), then we will use the existing file,
+        # otherwise, we will compress and upload the tarball.
+        
+        with tempfile.NamedTemporaryFile(suffix=".tar") as tar_fh:
+        
+            output_sha1 = hashlib.sha1()
+        
+            for dirname, subdirs, files in os.walk(resources_dir):
+                if not dirname.startswith(resources_dir):
+                    raise AssertionError('Expected %r to start with root directory %r' % (dirname, resources_dir))
+
+                # Add an entry for the directory itself
+                relative_dirname = dirname[len(resources_dir):]
+                dir_stat = os.lstat(dirname)
+                if not relative_dirname.startswith('/'):
+                    relative_dirname = '/' + relative_dirname
+
+                fields = [relative_dirname, str(dir_stat.st_mode), str(int(dir_stat.st_mtime * 1000))]
+                output_sha1.update(b''.join(s.encode('utf-8') + b'\0' for s in fields))
+                
+                # add an entry in the tar file for the current directory, but
+                # do not recurse!
+                subprocess.check_call(['tar', '-C', resources_dir, '-rf', tar_fh.name, '--no-recursion', '.' + relative_dirname])
```

Comment:
I'm concerned about the overhead of invoking tar here as well as below (once per directory and once per file being packed up). In the following benchmark, I tested building an app that contains the source distribution of node.js. A 19x slowdown is observed (though this is probably a pathologically large example, at 16,000 files and directories):

```
# master
psung@exabase ~/source/cat $ time dx build
[lint warnings stripped...]
real    0m32.804s
user    0m13.067s
sys     0m1.172s

# localSymlinkOnly
psung@exabase ~/source/cat $ time dx build
[lint warnings stripped...]
WARNING:dxpy:PUT <function get_upload_url_and_headers at 0x7f59308dcaa0>: ProtocolError: ('Connection aborted.', error(104, 'Connection reset by peer')). Waiting 1 seconds before retry 1 of 6...
INFO:dxpy:PUT <function get_upload_url_and_headers at 0x7f59308dcaa0>: Recovered after 1 retries
{"id": "applet-BvPPfp80BGv7g5Fx7JVqBzvb"}

real    10m38.361s
user    4m0.288s
sys     6m1.237s
```

Maybe @asimenos can point us to a large app? (Ideally, many small files) I'd be curious to see whether for a more realistic app the total time is dominated by the time to upload anyway.

Additionally, would it be possible to batch these calls so as to make `tar` process many files in a single command-line invocation?